From 4f76173d59058f87fc6c9da81818ce3330027a80 Mon Sep 17 00:00:00 2001
From: CI Assistant <yossi.ilkanaev@intel.com>
Date: Wed, 9 Jan 2019 05:23:55 +0100
Subject: [PATCH] Merge pull request #383 in SW_UGW/linux from
 bugfix/DRVLIB_SW-1103-fix-the-change-mtu-size-function-for-prx300-ethernet-driver
 to master

* commit '3c047686ca6cf69c0b4cfe269a6e74bbb15e421c':
  DRVLIB_SW-1103: fix the change MTU size function for PRX300 Ethernet driver
---
 drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c | 42 +++++++++++++++---------
 1 file changed, 26 insertions(+), 16 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
index b1966eb1a2a3..a10b52008274 100644
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -628,25 +628,35 @@ static int ltq_change_mtu(struct net_device *dev, int new_mtu)
 	if (new_mtu < ETH_ZLEN || new_mtu > LTQ_ETH_MAX_DATA_LEN)
 		return -EINVAL;
 
-	/* if the MTU > 1500, do the jumbo config in switch */
-	if (new_mtu > ETH_DATA_LEN && !(priv->jumbo_enabled)) {
-		if (priv->wan) {
-			if (ltq_enable_gsw_r_jumbo(dev) < 0)
-				return -EIO;
-		} else {
-			if (ltq_enable_gsw_l_jumbo(dev) < 0)
-				return -EIO;
+	if (g_eth_switch_mode == 0) {
+		if (priv->xgmac_id >= 0) {
+			struct mac_ops *ops;
+
+			ops = gsw_get_mac_ops(0, priv->xgmac_id);
+			if (ops)
+				ops->set_mtu(ops, new_mtu);
 		}
+	} else {
+		/* if the MTU > 1500, do the jumbo config in switch */
+		if (new_mtu > ETH_DATA_LEN && !(priv->jumbo_enabled)) {
+			if (priv->wan) {
+				if (ltq_enable_gsw_r_jumbo(dev) < 0)
+					return -EIO;
+			} else {
+				if (ltq_enable_gsw_l_jumbo(dev) < 0)
+					return -EIO;
+			}
 
-		ltq_enable_gsw_r_pmac_jumbo(dev);
-		priv->jumbo_enabled = 1;
-	} else if (priv->jumbo_enabled) {
-		if (priv->wan)
-			ltq_disable_gsw_r_jumbo(dev);
-		else
-			ltq_disable_gsw_l_jumbo(dev);
+			ltq_enable_gsw_r_pmac_jumbo(dev);
+			priv->jumbo_enabled = 1;
+		} else if (priv->jumbo_enabled) {
+			if (priv->wan)
+				ltq_disable_gsw_r_jumbo(dev);
+			else
+				ltq_disable_gsw_l_jumbo(dev);
 
-		priv->jumbo_enabled = 0;
+			priv->jumbo_enabled = 0;
+		}
 	}
 
 	dev->mtu = new_mtu;
